GSClient

benjamincerigo opened the pull request:
Implement of the GSClient:
- `get`,`put` and `remove` methods
- support for `gs` and `gcs` schemas
- unit tests
- functional test

Most of the implementation follows the `s3_client` apart from some small aspects that I have diverged from.

### Exceptions
In the `s3_client` there is an `_isfile` guard. I felt this was not necessary as the storage client will raise a `NotFound` if the object is not found.

In general, I let the google storage client raise the exceptions as needed. 

### `gs` and `gcs` schemas
- `gs` usage example: [by google](https://cloud.google.com/storage/docs/gsutil)
- `gcs` usage: [kedro](https://kedro.readthedocs.io/en/stable/04_user_guide/08_advanced_io.html?highlight=gcs#checkpoint-configuration), [fsspec](https://github.com/intake/filesystem_spec#purpose)

### unit test
The unit test separates check on that actual call on the google storage client and tests for the wrapping functions. 

The checks on the client, `test_gs_client.py:test_helper_<[get,put,remove]>`, are brittle but they that the correct call is made on the client. 

The check on the wrapping functions are the rest of the tests in `test_gs_client.py` and check that the guard method `_fetch_bucket_and_key` is working correctly for each wrapper function. 

### functional test
Since Google has no equivalent of the moto package for AWS I decide to write a functional test that interface with GCS directly. Not sure if this is in keeping what you expect from functional tests.

It is simple to mock the google storage client so let me know what you want.

### Configuration of the client
In the documentation of google storage client, they advise on using environment variables to configure the credentials and project of the client. As such, I did not implement `username:password` URL for the GSClient as this seems to go against the method they advise. 

I added links to the google docs in the docstring of the GSClient to direct troubleshooting developers. Let me know if documentation about this is needed.

capitancambio commented on src/tentaclio/clients/gs_client.py:
```diff
@@ -99,14 +99,14 @@ def _fetch_bucket_and_key(
 
-    def _get_bucket(self, bucket_name: str) -> storage.Bucket:
+    def _get_blob(self, bucket_name: str, key_name: str) -> storage.Bucket:
         """Get the bucket client."""
```
:panda_face: the comment should read _blob_

capitancambio commented:
Oh just saw the comment. I'd rather not have to configure connections to aws/gc to test this as it's always a pita to have it well configured in all the environments. So if you can mock it would be great. Otherwise I don't think it would be a blocker,  we probably wouldn't have functional tests for aws if it weren't for moto.

ali-tny commented:
I've had a look before, and there isn't an equivalent of moto for google cloud storage ([and it's unlikely there'll be an official one](https://github.com/googleapis/google-cloud-python/issues/5816). Apparently there is for the java client, but not for python

capitancambio commented:
it's fine then. I never was a huge fan of functional tests with something that is not the real deal. At the same time I'm not a huge fan of having to configure test accounts for these services and break everyone's testing environment.

capitancambio commented:
so @benjamincerigo just get rid of the functional tests for gc and we're ready to go

benjamincerigo commented:
So I have added:
-  a `skipif` that depends on the `TENTACLIO__CONN__GS_TEST` for the functional test that makes the real calls
- a test that checks that the API calls reaches the mocks for the schemas `gs` and `gcs`

I thought that it was good to have the mocked version so that the intergration between the API and the `GSClient` is tested and I like being able to run the real thing if I want to.

PS. I added it as a `--fixup` commit thus before merge run `git rebase -i master --autosquash` to fix up the commit in the correct place. (Always nice to have a clean history).

capitancambio commented:
@benjamincerigo thanks a million!

benjamincerigo commented:
Awesome. 

Thanks a bunch for reviewing my work. I am super stoked that you guys are going to merge it! 

Let me know if I can help with anything else.

capitancambio commented:
oh, thanks for merging @ali-tny I got distracted with other things, I'll prepare a new release including this so it's pypy installable

benjamincerigo commented:
@ali-tny and @capitancambio. Thanks a bunch, guys. Was super fun to work on this project!

PS. Not sure if there should also be an update to the README.md file once the release is done.
